src/realm.js

```diff
@@ -453,6 +461,99 @@ export class Realm {
+      }
+      invariant(resultVal instanceof Value);
+      return resultVal;
+    } catch (e) {
+      if (diagnostic !== undefined) return diagnostic;
+      throw e;
+    } finally {
+      this.errorHandler = savedHandler;
+    }
+  }
+
+  evaluateForFixpointEffects(
+    loopContinueTest: () => Value,
+    loopBody: () => EvaluationResult
+  ): void | [Effects, Effects] {
+    try {
+      let effects1 = this.evaluateForEffects((loopBody: any));
+      while (true) {
+        this.restoreBindings(effects1[2]);
+        this.restoreProperties(effects1[3]);
+        let effects2 = this.evaluateForEffects(() => {
+          let test = loopContinueTest();
+          if (!(test instanceof AbstractValue)) throw new FatalError("loop terminates before fixed point");
+          return (loopBody(): any);
+        });
+        this.restoreBindings(effects1[2]);
+        this.restoreProperties(effects1[3]);
+        if (Widen.containsEffects(effects1, effects2)) {
+          // effects1 includes every value present in effects2, so doing another iteration using effects2 will not
+          // result in any more values being added to abstract domains and hence a fixpoint has been reached.
+          let [, , bindings1, pbindings1] = effects1;
+          if (pbindings1.size > 0) return undefined;
+          let [, gen, bindings2, pbindings2] = effects2;
+          if (pbindings2.size > 0) return undefined;
+          this._emitLocalAssignments(gen, bindings1, bindings2);
+          return [effects1, effects2];
+        }
+        effects1 = Widen.widenEffects(this, effects1, effects2);
+      }
+    } catch (e) {
+      return undefined;
+    }
+  }
+
+  // populate the loop body generator with assignments that will update the phiNodes
+  _emitLocalAssignments(gen: Generator, bindings1: Bindings, bindings2: Bindings) {
+    // bindings1 maps local bindings to widened values whose build nodes return the identity of the correspoding phiNodes
+    // bindings2 maps local bindings to the unwidened values whose build nodes result in expressions that reference phiNodes
+    let idFor: Map<any, BabelNodeIdentifier> = new Map();
+    bindings1.forEach((val, key, map) => {
+      if (val instanceof AbstractValue && val.kind === "widening") {
+        let id = val.buildNode([]);
+        idFor.set(key, (id: any));
+      }
+    });
+    bindings2.forEach((val, key, map) => {
+      if (val instanceof AbstractValue) {
+        invariant(val._buildNode !== undefined);
+        let tval = gen.derive(val.types, val.values, val.args, val._buildNode, {
+          skipInvariant: true,
+        });
+        let id = idFor.get(key);
+        invariant(val !== undefined);
+        gen.emitStatement([tval], ([v]) => {
+          invariant(id !== undefined);
+          invariant(val instanceof AbstractValue);
+          val._buildNode = nodes => (id: any);
+          val.kind = "widened";
+          return t.expressionStatement(t.assignmentExpression("=", id, v));
```

Comment:
All assignments are emitted in some order. Couldn't the right-hand sides refer to the left-hand sides in some cyclic way, resulting in bad code generation? (I tried playing around for a while, but I couldn't produce a test case that exposes this...)

Reply:
I'm a bit worried about this too. Let me try and cook up a test case that will clarify things one way or another.

Reply:
Got a test case that showed up some problems with sequencing. Things have now been refactored so that every rh side is first evaluated into temporary variables before the locals (in the guise of their phi nodes) are updated. Any references to phi nodes inside the loop body will always be to their values at loop iteration start. I.e. everything is done the SSA way until all expressions have been evaluated, then the phi nodes are updated in a batch of assignments that do no expression evaluation.